src/safeposix/vmmap.rs

```diff
@@ -0,0 +1,587 @@
+use super::vmmap_constants::*;
+use std::io;
+use nodit::NoditMap;
+use nodit::{interval::ie, Interval};
+
+/// Used to identify whether the vmmap entry is backed anonymously,
+/// by an fd, or by a shared memory segment
+
+#[allow(dead_code)]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum MemoryBackingType {
+    None, // just a dummy value for places where it needs to be passed, but you dont have the value
+    Anonymous,
+    SharedMemory(u64),   // stores shmid
+    FileDescriptor(u64), // stores file descriptor addr
+}
+
+/// In the old native client based vmmap, we relied on the fd, shmid
+/// fields. 
+/// 
+/// An entry in the virtual memory map that contains fields such as page number, number of pages, 
+/// permissions, file offset, file size, shared memory ID, and backing fields to distinguish memory types.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct VmmapEntry {
+    pub page_num: u32, /* base virtual addr >> NACL_PAGESHIFT */
+    pub npages: u32,   /* number of pages */
+    pub prot: i32,     /* mprotect attribute */
+    pub maxprot: i32,
+    pub flags: i32,       /* mapping flags */
+    pub removed: bool,    /* flag set in fn Update(); */
+    pub file_offset: i64, /* offset into desc */
+    pub file_size: i64,   /* backing store size */
+    pub cage_id: u64,
```

Comment:
I think a note here for why we store the cageid here would be good. I believe its because we need it for when we need to check file protections for a memory backing.

Reply:
I suppose this is fine for now, but in the future, it may be that there are multiple cage_ids able to simultaneously access a part of memory.   Perhaps add a comment indicating this may need to be rethought later, unless there is a simple structural change which would make sense in this case...

Reply:
This is the cage_id to refer to for file backings. That probably should be noted like that.

If we wanted multiple cages to access the same memory, we'd probably have to keep the cage id per file backed memory, though doing something like that would probably require rethinking the whole vmmap.